java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/PriorityQueueRun.java

```diff
@@ -0,0 +1,199 @@
+
+        for (Entry<Priority, List<TaskDescription<?, ?>>> e : priority2Tasks.entrySet()) {
+            if (!e.getValue().isEmpty()) {
+                foundPriority = e.getKey();
+                foundTasks = e.getValue();
+                break;
+            }
+        }
+
+        if (foundPriority == null) {
+            return ;
+        }
+
+        if (currentPriority == null) {
+            Priority thisPriority = currentPriority = foundPriority;
+            TaskDescription thisTask = foundTasks.remove(0);
+            CancelCheck thisTaskCheck = currentTaskCheck = new CancelCheck();
+
+            WORKER.post(() -> {
+                thisTask.result.whenComplete((__, exc) -> {
+                    if (exc instanceof CancellationException) {
+                        thisTaskCheck.cancel();
+                    }
+                });
+
+                if (thisTask.result.isCancelled()) {
+                    //nothing to do anymore:
+                    return ;
+                }
+
+                Object result = null;
+                Throwable exception = null;
+
+                try {
+                    result = thisTask.task.compute(thisTask.data, thisTaskCheck);
+                } catch (Throwable t) {
+                    exception = t;
+                }
+
+                synchronized (PriorityQueueRun.this) {
+                    currentPriority = null;
+                    currentTaskCheck = null;
+
+                    if (!thisTask.result.isCancelled()) {
+                        if (exception != null) {
+                            thisTask.result.completeExceptionally(exception);
+                        } else if (thisTaskCheck.isCancelled()) {
+                            priority2Tasks.computeIfAbsent(thisPriority, __ -> new ArrayList<>())
+                                          .add(0, thisTask);
+                        } else {
+                            thisTask.result.complete(result);
+                        }
+                    }
+                }
+
+                scheduleNext();
+            });
+        }
+
+        if (currentPriority != null && currentPriority.compareTo(foundPriority) < 0) {
+            //cancel the currently running task:
+            currentTaskCheck.cancel();
+        }
+    }
+
+    private static final class TaskDescription<P, R> {
+        private final CancellableTask<P, R> task;
+        private final P data;
+        private final CompletableFuture<R> result;
+
+        public TaskDescription(CancellableTask<P, R> task, P data, CompletableFuture<R> result) {
+            this.task = task;
+            this.data = data;
+            this.result = result;
+        }
+    }
+
+    public interface CancellableTask<P, R> {
+        public R compute(P param, CancelCheck cancel) throws Exception;
+    }
+
+    public interface CancelCallback {
+        public void cancel();
+    }
+
+    public final class CancelCheck {
```

Comment:
Is this supposed to be a public class? It is currently used only within this class.